tarodb/postgres.go

```diff
@@ -0,0 +1,143 @@
+	DefaultPostgresFixtureLifetime = 10 * time.Minute
+)
+
+// PostgresConfig holds the postgres database configuration.
+type PostgresConfig struct {
+	SkipMigrations     bool   `long:"skipmigrations" description:"Skip applying migrations on startup."`
+	Host               string `long:"host" description:"Database server hostname."`
+	Port               int    `long:"port" description:"Database server port."`
+	User               string `long:"user" description:"Database user."`
+	Password           string `long:"password" description:"Database user's password."`
+	DBName             string `long:"dbname" description:"Database name to use."`
+	MaxOpenConnections int32  `long:"maxconnections" description:"Max open connections to keep alive to the database server."`
+	RequireSSL         bool   `long:"requiressl" description:"Whether to require using SSL (mode: require) when connecting to the server."`
+}
+
+// DSN returns the dns to connect to the database.
+func (s *PostgresConfig) DSN(hidePassword bool) string {
+	var sslMode = "disable"
+	if s.RequireSSL {
+		sslMode = "require"
+	}
+
+	password := s.Password
+	if hidePassword {
+		// Placeholder used for logging the DSN safely.
+		password = "****"
+	}
+
+	return fmt.Sprintf(dsnTemplate, s.User, password, s.Host, s.Port,
+		s.DBName, sslMode)
+}
+
+// PostgresStore is a database store implementation that uses a Postgres
+// backend.
+type PostgresStore struct {
+	cfg *PostgresConfig
+
+	DB *sql.DB
+
+	*sqlc.Queries
+}
+
+// NewPostgresStore creates a new store that is backed by a Postgres database
+// backend.
+func NewPostgresStore(cfg *PostgresConfig) (*PostgresStore, error) {
+	log.Infof("Using SQL database '%s'", cfg.DSN(true))
+
+	rawDb, err := sql.Open("pgx", cfg.DSN(false))
+	if err != nil {
+		return nil, err
+	}
+
+	if !cfg.SkipMigrations {
+		// Now that the database is open, populate the database with
+		// our set of schemas based on our embedded in-memory file
+		// system.
+		//
+		// First, we'll need to open up a new migration instance for
+		// our current target database: sqlite.
+		driver, err := postgres_migrate.WithInstance(
+			rawDb, &postgres_migrate.Config{},
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		postgresFS := newReplacerFS(sqlSchemas, map[string]string{
+			"BLOB":                "BYTEA",
+			"INTEGER PRIMARY KEY": "SERIAL PRIMARY KEY",
+			"TIMESTAMP":           "TIMESTAMP WITHOUT TIME ZONE",
+		})
+
+		err = applyMigrations(
+			postgresFS, driver, "sqlc/migrations", cfg.DBName,
+		)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	queries := sqlc.New(rawDb)
+
+	return &PostgresStore{
+		DB:      rawDb,
+		cfg:     cfg,
+		Queries: queries,
+	}, nil
+}
+
+// BeginTx wraps the normal sql specific BeginTx method with the TxOptions
+// interface. This interface is then mapped to the concrete sql tx options
+// struct.
+func (s *PostgresStore) BeginTx(ctx context.Context, opts TxOptions) (*sql.Tx,
```

Comment:
This is identical for the `SqliteStore` and `PostgresStore` structs, perhaps we can move into some struct (w/ the base functionality) then embed here.

Reply:
Cool, did that :+1: